cpp/include/raft/selection/col_wise_sort.cuh

```diff
@@ -0,0 +1,346 @@
+        }
+
+// helper function to layout offsets for rows for DeviceSegmentedRadixSort
+        template <typename T>
+        cudaError_t layoutSortOffset(T* in, T value, int n_times, cudaStream_t stream)
+        {
+            dim3 block(128);
+            dim3 grid((n_times + block.x - 1) / block.x);
+            devOffsetKernel<T><<<grid, block, 0, stream>>>(in, value, n_times);
+            return cudaGetLastError();
+        }
+
+/**
+ * @brief sort columns within each row of row-major input matrix and return sorted indexes
+ * modelled as key-value sort with key being input matrix and value being index of values
+ * @param in: input matrix
+ * @param out: output value(index) matrix
+ * @param n_rows: number rows of input matrix
+ * @param n_cols: number columns of input matrix
+ * @param bAllocWorkspace: check returned value, if true allocate workspace passed in workspaceSize
+ * @param workspacePtr: pointer to workspace memory
+ * @param workspaceSize: Size of workspace to be allocated
+ * @param stream: cuda stream to execute prim on
+ * @param sortedKeys: Optional, output matrix for sorted keys (input)
+ */
+        template <typename InType, typename OutType>
+        void sortColumnsPerRow(const InType* in,
+                               OutType* out,
+                               int n_rows,
+                               int n_columns,
+                               bool& bAllocWorkspace,
+                               void* workspacePtr,
+                               size_t& workspaceSize,
+                               cudaStream_t stream,
+                               InType* sortedKeys = nullptr)
+        {
+            // assume non-square row-major matrices
+            // current use-case: KNN, trustworthiness scores
+            // output : either sorted indices or sorted indices and input values
+            // future : this prim can be modified to be more generic and serve as a way to sort column entries
+            // per row
+            //          i.e. another output format: sorted values only
+
+            int totalElements          = n_rows * n_columns;
+            size_t perElementSmemUsage = sizeof(InType) + sizeof(OutType);
+            size_t memAlignWidth       = 256;
+
+            // @ToDo: Figure out dynamic shared memory for block sort kernel - better for volta and beyond
+            // int currDevice = 0, smemLimit = 0;
+            // CUDA_CHECK(cudaGetDevice(&currDevice));
+            // CUDA_CHECK(cudaDeviceGetAttribute(&smemLimit, cudaDevAttrMaxSharedMemoryPerBlock, currDevice));
+            // size_t maxElementsForBlockSort = smemLimit / perElementSmemUsage;
+
+            // for 48KB smem/block, can fit in 6144 4byte key-value pair
+            // assuming key-value sort for now - smem computation will change for value only sort
+            // dtype being size of key-value pair
+            std::map<size_t, int> dtypeToColumnMap = {{4, 12288},   // short + short
```

Comment:
It's not immediately obvious to me how often this will be called, but if performance is a concern, we should make this a switch statement (or if we expect the number of cases to grow, a sorted vector) rather than using a `std::map`. For small maps like this, the cache-unfriendly accesses in the underlying RB-tree add up surprisingly quickly. In Treelite, we had a very similar `std::map` usage that didn't seem like it would be a problem, but profiling showed that those map accesses added a shocking amount to runtime.